[NHUB-533] Implement GlobalTopics async

eos87 opened the pull request:
### Purpose
Implement the GlobalTopics management with new Async framework

### What has changed
- Added new model, service class and module registration
- Updated views to use new endpoints group + reference to new service class

This PR depends on https://github.com/superdesk/superdesk-core/pull/2688

Resolves: [NHUB-533]

[NHUB-533]: https://sofab.atlassian.net/browse/NHUB-533?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

eos87 committed:
Update upload-artifact to v4

eos87 committed:
Implement `navigations` async

NHUB-533

eos87 committed:
Update references and fix tests

NHUB-533

eos87 committed:
Minor fix to model

NHUB-533

eos87 committed:
Use request.abort instead of flask abort

NHUB-533

eos87 committed:
Fix tests temporarily

eos87 committed:
Small improvement to views

NHUB-533

eos87 committed:
Fix e2e tests

NHUB-533

eos87 committed:
Fix tests

NHUB-533

eos87 committed:
Skip flake8 issue

NHUB-533

MarkLark86 commented:
@eos87 Some WebAPI behave tests are failing. One of those is due to using [get_resource_service(resource).delete_action()](https://github.com/superdesk/superdesk-core/blob/async/superdesk/tests/steps.py#L503) from superdesk-core. This could be fixed by attempting to get the resource from the new async framework first, i.e.

```python
from superdesk.core import get_current_async_app
from superdesk import get_resource_service

async def run_delete_action_for_resource(resource: str) -> None:
    app = get_current_async_app()
    try:
        await app.resources.get_resource_service(resource).delete_many({})
    except KeyError:
        get_resource_service(resource).delete_action()
```

You will then need to do something similar with [get_resource_service(resource).post(items)](https://github.com/superdesk/superdesk-core/blob/async/superdesk/tests/steps.py#L510) to attempt to use the async resource, falling back to legacy sync resource.

MarkLark86 commented on e2e/cypress/e2e/wire/topics.cy.js:
```diff
@@ -98,10 +98,15 @@ describe('Wire - Topics', function () {
-            .getTopicCardAction('Sofab Weather', 'Remove from folder')
-            .should('exist');
+
+        // TODO-ASYNC: The dragTopicToFolder action triggers a PATCH action over a nested resource.
```
This should be resolved once @devketanpro has finished with his [Topic Folder PR](https://github.com/superdesk/newsroom-core/pull/1059)

MarkLark86 commented on e2e/server/newsroom_e2e/module.py:
```diff
@@ -45,7 +70,13 @@ async def populate_resources(request: Request) -> Response:
             items = entry.get("items") or []
 
-            if entry.get("use_resource_service", True):
+            if entry.get("use_db_collection", False):
```
This might not be ideal. The idea behind E2E is to reproduce as close as possible a production environment with testing user interactions. Here we aren't using the resource service, which might have side effects to creating resources. Which would mean we won't need custom serialisation here but use the one the rest of the system uses.

Instead might I suggest to continue to use the `use_resource_service` option, and attempt to get the resource service with `app.resources.get_resource_service(resource)` from the async framework, falling back to the legacy sync resources.

eos87 replied:
Improvement done, please check when you have some time.

MarkLark86 commented on newsroom/navigations/module.py:
```diff
@@ -0,0 +1,38 @@
+from .views import get_settings_data, navigations_endpoints
+from .service import NavigationsService
+
+if TYPE_CHECKING:
```
What is the reason to only include this when `TYPE_CHECKING` is enable?

eos87 replied:
a circular reference ~but I will double check this because I have changed many pieces that this might not be needed~.

MarkLark86 commented on newsroom/navigations/model.py:
```diff
@@ -0,0 +1,11 @@
+
+
+class Navigation(NewshubResourceModel):
+    name: str
```
This value should be unique, using [validate_iunique_value_async](https://github.com/superdesk/newsroom-core/blob/async/newsroom/users/model.py#L40C45-L40C73)

MarkLark86 commented on newsroom/products/views.py:
```diff
@@ -67,6 +69,13 @@ def validate_product(product):
         return jsonify({"name": gettext("Name not found")}), 400
 
 
+async def find_nav_or_404(id: str) -> Navigation:
```
[id](https://docs.python.org/3/library/functions.html#id) is a built-in python function, so there is a name clash here. Would be good to use a different name instead

MarkLark86 commented on tests/core/utils.py:
```diff
@@ -1,7 +1,40 @@
+        collection_name (str): The name of the MongoDB collection.
+        data (list[Any]): A list of documents to insert into the collection.
+    """
+    db_collection = get_db_collection(collection_name)
```
Same with this one, it would be better to use the `app.resources.get_resource_service`. Not only does this apply side effects, serialisation, but also makes sure the items are created in Elastic (if enabled for the resource)

MarkLark86 commented on newsroom/agenda/views.py:
```diff
@@ -15,7 +15,7 @@
 from newsroom.template_filters import is_admin_or_internal, is_admin
 from newsroom.topics import get_company_folders, get_user_folders, get_user_topics
-from newsroom.navigations.navigations import get_navigations
+from newsroom.navigations.utils import get_navigations
```
Would it be possible to add the following to the `__all__` attribute in `newsroom.navigations.__init__.py`:
- get_navigations
- get_navigation_by_company
- get_navigations_as_list

This way we can define what the `newsroom.navigations` module exposes, and not need to import deeper into that module

`from newsroom.navigations.utils import get_navigations`
would then become
`from newsroom.navigations import get_navigations`

MarkLark86 commented on newsroom/cards/module.py:
```diff
@@ -26,11 +27,14 @@
 async def get_settings_data():
     app = get_current_wsgi_app()
 
+    cursor = await NavigationsService().search(lookup={"is_enabled": True})
```
In situations like this, where we have multiple `await` for separate DB requests, we can probably start using `asyncio.gather` to run both requests concurrently and await it's results.

In this case, it would be:
```python
from asyncio import gather

async def get_settings_data():
    ...
    cursors = await gather(
        NavigationsService().search({"is_enabled": True}),
        CardsResourceService().search({}),
    )
    [navigations, cards] = await gather(
        cursors[0].to_list_raw(),
        cursors[1].to_list_raw(),
    )

    return {
        ...,
        "cards": cards,
        "navigations": navigations,
    }
```

eos87 replied:
good idea. I've made the changes.

MarkLark86 commented on newsroom/navigations/views.py:
```diff
@@ -1,77 +1,123 @@
-    set_original_creator(nav_data)
-    ids = get_resource_service("navigations").post([nav_data])
+    creation_data = await prepare_navigation_data(nav_data)
+    creation_data["id"] = service.generate_id()
```
We could probably have this in the base async service create method. If supplied item is a dict and `_id` not defined, assign using the `self.generate_id()` method.

This way we dont' have to worry about it with these create item routes

eos87 replied:
yeah, good idea. I'll do that in a separate PR to superdesk-core and then another one for newsroom core.

eos87 committed:
Implement some feedback from review

NHUB-533

eos87 committed:
Set superdesk-core dependency back to async

NHUB-533

eos87 committed:
Fix some of the behave tests

NHUB-533

eos87 committed:
Fix behave tests

NHUB-533

eos87 committed:
Make Navigation name unique

NHUB-533

eos87 committed:
Update validators.py

eos87 committed:
Remove not used import

eos87 committed:
Merge remote-tracking branch 'origin/async' into hg/NHUB-533-global-topics-async

eos87 committed:
Use resource service instead of inject into db directly

Implement improvements for the E2E tests and unit tests to use resource service instead of injecting directly into the database collection.

NHUB-533

MarkLark86 commented on newsroom/cards/module.py:
```diff
@@ -26,11 +27,16 @@
+    cards_task = CardsResourceService().find({})
+    navs_task = NavigationsService().search(lookup={"is_enabled": True})
+
+    cards, navigations = await gather((await cards_task).to_list_raw(), (await navs_task).to_list_raw())
```
Using `await` is a synchronous action, meaning this will await `cards_task` first, then await `navs_task` sequentially. To gain the benefit of concurrently awaiting both requests for a cursor, we will have to use something like:
 ```python
# Await the request for a cursor
[cards_cursor, navs_cursor] = await gather(cards_task, navs_task)
# Await for the cursor data
[cards, navigations] = await gather(cards_cursor.to_list_raw(), navs_cursor.to_list_raw())
```

This is because the initial request gets sent to MongoDB, which returns a cursor to the request. This is an async I/O request.

Then `to_list_raw` then requests MongoDB to return items from the cursor. This is another async I/O request.

So because we have 2 separate I/O requests to MongoDB, we will need to gather twice.

MarkLark86 replied:
Sorry, I got my information wrong here. Your changes here should suffice for mongo resources.

**Mongo:**
mongo.find: sync
cursor: async

**Elastic**
elastic.search: async
cursor: sync

MarkLark86 commented on newsroom/core/resources/validators.py:
```diff
@@ -62,3 +64,29 @@ def _validate_dashboard_type_exists(value: str | None = None) -> str | None:
+
+    async def _validate_iunique_value(item: ResourceModel, value: Any) -> None:
+        fields = {field: getattr(item, field) for field in field_names}
+        if any(value for value in fields.values()):
```
What is the purpose of this if/return statement?

eos87 replied:
my bad, I changed this last moment, it should have been 
```python
if any(value is None for value in fields.values()):
```
I'll fix it now.

MarkLark86 reviewed (commented):
Looks good, just a couple small things then would be good to merge

eos87 committed:
Fix validator and add unit tests

NHUB-533

eos87 committed:
Remove decorator as it is not needed

NHUB-533
